Adds the Rotatium chemical [DNM]

optimumtact opened the pull request:
Adds Rotatium, a chemical that causes the person who consumes it to
rotate their view, causing distorted shit like the proc would do.

Ported from HippieStation
https://github.com/HippieStationCode/HippieStation13/pull/1275

Thanks to TDcoolguy300

balance/non code feedback concerns that are greater than a thumbs up / down go here please
https://tgstation13.org/phpBB/viewtopic.php?f=10&t=5276

:cl: oranges
experiment: Nanotrasen chemists announced today a new breakthrough in the field of mind altering substances, the new drug is expected to enter fullscale trials on research stations in the near future. Initial trial results have been described as "promising"
rscadd: Ports Rotatium, a drug that causes your view to rotate funkily, from Hippie. The effects of rotatium stop after it's gone under 2 units, or is purged by Calomel, and your view is restored to normal.
/:cl:

TDcoolguy300 committed:
Adds the Rotatium chemical

Adds Rotatium, a chemical that causes the person who consumes it to
rotate their view, causing distorted shit like the proc would do.

Ported from HippieStation
https://github.com/HippieStationCode/HippieStation13/pull/1275

Conflicts:
	code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm
	code/modules/reagents/Chemistry-Recipes/Toxins.dm

RemieRichards commented on code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm:
```diff
@@ -642,6 +642,37 @@
+	if(rotate_timer >= rand(5,30)) //Random rotations are wildly unpredictable and hilarious
+		rotate_timer = 0
+		hold = rand(1,8)
+		switch(hold)
```
so it's a 1 in 4 to pick 1 from a list of 4...

why is this not ```M.client.dir = pick(2,4,6,8)```

also 6 is SOUTHEAST, and I dont think that actually works for ```client.dir```

optimumtact replied:
I'll fix this up

So only N, S, W, E rotations will work for client dir?

RemieRichards replied:
Yep!

AnturK commented on code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm:
```diff
@@ -642,6 +642,37 @@
+/datum/reagent/toxin/rotatium/on_mob_life(mob/living/M)
+	rotate_timer++
+	if(M.reagents.get_reagent_amount("rotatium") < 2)
+		M.client.dir = 1
```
Extra edge case but wouldn't purging make it stick ?

optimumtact replied:
Handled! https://github.com/optimumtact/-tg-station/blob/rotatiuminim/code/modules/reagents/Chemistry-Reagents/Medicine-Reagents.dm#L345

AnturK replied:
I'm blind. All good.

RemieRichards replied:
@optimumtact can you put this on ```on_mob_delete()``` too?
otherwise people will be stuck with a weird client dir forever in cases where it wasn't removed by camomel, such as admin force removals.

optimumtact replied:
Can do

optimumtact replied:
done

RemieRichards replied:
now to be a tiny bit nitpicky (it'll also remove these line comments), replace ```1``` with ```NORTH```

optimumtact committed:
Make some improvements to the rotatium chemical

Improves rotation logic to use a pick list of directions
Resets client dir on chem deletion from mob
Makes all directions to their appropriate DEFINES

phil235 commented on code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm:
```diff
@@ -642,6 +642,37 @@
+	toxpwr = 0
+	var/rotate_timer = 0
+	var/randrot = 1
+	var/hold = 0
```
why does hold need to be a var?

phil235 commented on code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm:
```diff
@@ -642,6 +642,37 @@
+				randrot = 6
+			if(7 to 8)
+				randrot = 8
+		M.client.dir = randrot
```
can't you do `M.client.dir = pick(2,4,6,8)` ?

optimumtact commented:
Thanks for the reviews!

TDcoolguy300 commented:
As you can tell my methods can be very strange, I'll be sure to quickly bring over the new rotation method. Thanks for cleaning this up Optimumtact.

TDcoolguy300 commented on code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm:
```diff
@@ -642,6 +642,30 @@
+/datum/reagent/toxin/rotatium/on_mob_life(mob/living/M)
+	rotate_timer++
+	if(M.reagents.get_reagent_amount("rotatium") < 2)
+		M.client.dir = NORTH
```
I highly recommend adding a ..() after the direction correction. With the return there the ..() below will not actually do anything

optimumtact replied:
good catch

Iamgoofball commented:
Holy shit. That's actually really cool.

:+1:

PKPenguin321 commented:
yeah this is actually super neat. :+1:

Wyzack commented:
>Oranges makes a PR that is not a meme change to retardspeak

i like it

optimumtact commented:
it's oranges

Cheridan commented:
what does mindbreaker smell like

optimumtact commented:
Have you ever had an iranian curry?

MrPerson commented:
I thought there were still a ridiculous number of bugs related to client.dir changing including, but not limited to:
Changing keybindings for northeast, southeast, etc
Graphical glitches with things that need to look a certain way next to other things (smoothing?)
Problems with Click() args, although there's a chance byond does these correctly

No, having bugs like these are not "part of the experience" or whatever bullshit dumb people will spout out, they need to be fixed.

WJohn commented:
This should totally make the person who consumed it spin constantly.

optimumtact commented:
There were not any bugs in my play testing of this.

@WJohn I'll implement that instead if any one finds any unfix-able bugs

Iamgoofball commented:
@optimumtact make sure to add a subvariant called fliptonium for just spinning the person

Aranclanos commented on code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm:
```diff
@@ -642,6 +642,29 @@
+
+/datum/reagent/toxin/rotatium/on_mob_life(mob/living/M)
+	rotate_timer++
+	if(M.reagents.get_reagent_amount("rotatium") < 2)
```
doesnt the src contains the amount var

phil235 replied:
Aran is right. Use the "volume" var instead.

YotaXP commented:
- What happens if you ghost while your view is flip turned upside down?
- What happens if the chem wears off while on one of those shuttles that rotates your view?

RemieRichards commented:
@YotaXP shuttles don't rotate your view, they rotate the physical world and your character, you're still facing NORTH.

MrStonedOne commented:
ya, shuttle rotation rotates the actual shuttle, relocating the turfs, changing the direction of icons, and re-mounting wall mounts/rotating pixelx/y

optimumtact commented:
@YotaXP I'll test ghosting.

YotaXP commented:
Test ghosting, and other mind transfering stuff. (Like debraining)

xxalpha commented:
Post pictures.

Bawhoppen commented:
@MrPerson just because you say "it isn't" doesn't invalidate the argument

optimumtact commented:
This isnt' a problem right?
![just fuck my shit up](http://puu.sh/lCuno/cf2b4fcfeb.png)

Bawhoppen commented:
dont use rotatium if you dont want to hallucinate

duncathan commented:
Don't see any issues here

MrPerson commented:
So do you plan on fixing those issues or can we just go ahead and close this?

optimumtact commented:
I'm not sure what issues you are referring to?

MrStonedOne commented:
You need to add code to mob/living/logout to reset the view, but only if it was changed by this chem

You need to add null checks to everywhere you access the client var of a mob to ensure it's not null to prevent runtimes,

MrPerson commented:
@optimumtact
Listen here you cheeky little cunt. Don't give me that bullshit. Fix your goddamn bugs or fuck off.

optimumtact commented:
Excuse me officer, that's not a bug it's my artistic vision

MrPerson commented:
Your artistic vision is shit

Iamgoofball commented:
Is it even possible to fix that sprite shit? Can we get some sprite proc gurus in here to brainstorm a fix?

Razharas commented:
What about mob death?
The cient will stay turned?

Razharas commented on code/modules/reagents/Chemistry-Reagents/Toxin-Reagents.dm:
```diff
@@ -642,6 +642,29 @@
+	var/rotate_timer = 0
+
+/datum/reagent/toxin/rotatium/on_mob_life(mob/living/M)
+	rotate_timer++
```
````
if(!(M.client))
    return

Razharas commented on code/modules/reagents/Chemistry-Reagents/Medicine-Reagents.dm:
```diff
@@ -346,6 +346,7 @@
 	for(var/datum/reagent/R in M.reagents.reagent_list)
 		if(R != src)
 			M.reagents.remove_reagent(R.id,2.5)
+			M.client.dir = NORTH
```
Id prefer it just removed it like everything else so you dont need special client presence check

optimumtact replied:
?

Razharas replied:
Instead of explicitly returning the client dir to normal each life just make it purge the riotatium as usual and make the dir return only on removed() of rotatium(or how is this proc called when chem lefts the body i forgot)

optimumtact replied:
ahah, I see what you mean

optimumtact commented:
The next person who asks me to check for a null client is gonna get a bullet I swear on my life.

Dorsisdwarf commented:
Oranges oranges oranges

optimumtact commented:
It's unlikely this will be finished in it's current form as client.dir just manifests an extremely large number of bugs in our code.

Iamgoofball commented:
are they purely visual or are there game related bugs too?

Bawhoppen commented:
I imagine that it would not work for something like the painting canvas

optimumtact commented:
from my testing they seemed mostly visual bugs

xxalpha commented:
If it's visual you can just say it's part of the drug's effects.

Razharas commented:
Fix the runtimes with null client

optimumtact commented:
Yes thank you @Razharas I am aware of them

Razharas commented:
><oranges> rotatium kinda dea

xxalpha commented:
What was wrong with this again?

optimumtact commented:
There are a lot of sprites that don't have the appropriate directions, so they remain fixed when everything else rotates.

nullbear commented:
well i mean you can add them as-needed, on a prioritized basis. Ie. if a player notices that X looks really weird under the influence of rotatium, then mark X for additional icon dir_states to be added.

KorPhaeron commented:
stale

xxalpha commented:
MEME'D
